### PR TITLE
Feat/18889/update buttonlink ts

### LINKED ISSUE #18889: Migrate components to TS: ButtonLink
### Description
Migrate `ButtonLink` component to TypeScript

### Technical Details
- Convert Objects in `SNAKE_CASE` to enums in `PascalCase` follow [TS conventions](https://www.notion.so/Use-TypeScript-9ec7bd8fc8cf4d82b81cd8943c4707cd?pvs=4#7e6410b21100455a9a08442108f19d78)
- Update file name from `component-name.contants.js` to `component-name.types.ts` and add types
- Update file name from `component-name.js` to `component-name.tsx` and add types
- Update file name from `component-name.test.js` to `component-name.test.tsx`
- Update file name from `component-name.stories.js` to `component-name.stories.tsx`
- Update all instances of component objects to enums
- Update all documentation 

### Acceptance Criteria
- Component has been converted to TypeScript

### PR DESCRIPTION
**Pull request by garrettbear**
## Explanation

* Fixes #18889 
* Fixes #20214 

- Update ButtonLink to TS
- Fix ButtonLink alignment issue ( this issue came from padding being added my user style agent so I added 0px to the ButtonBase as a reset) 
- Fix ButtonLink hover underline to not apply when loading
- Fix ButtonLink loading color to be the same color

### Before
Screenshot 2023-08-02 at 3 50 58 PM
Screenshot 2023-08-02 at 3 51 30 PM

### After
Screenshot 2023-08-02 at 3 51 41 PM
Screenshot 2023-08-02 at 3 51 51 PM

## Manual Testing Steps
`yarn jest ui/components/component-library`

- Visit ButtonLink component on Storybook to confirm props work as expected.
- There should be no underline on hover when in loading state. 
- The color of the loading wheel should reflect the ButtonLink text color. 
- Make sure the ButtonLink text is aligned when ButtonLinkSize.Inherit is being used. 

## Pre-merge author checklist

- [X] I've clearly explained:
  - [X] What problem this PR is solving
  - [X] How this problem was solved
  - [X] How reviewers can test my changes
- [ ] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [X] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [X] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

**Review by darkwing — DISMISSED** [review]
Manual testing looks good, no changes seen.  Nice work!

File: ui/components/component-library/button-base/button-base.tsx [hunk 1]
```diff
@@ -58,6 +58,7 @@ export const ButtonBase: ButtonBaseComponent = React.forwardRef(
         {...(href && externalLink
           ? { target: '_blank', rel: 'noopener noreferrer' }
           : {})}
+        padding={0}
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/button-base/button-base.tsx:61]: I think using the Box props is good.

File: ui/components/component-library/button-link/button-link.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+  TextVariant,
+  TextColor,
+} from '../../../helpers/constants/design-system';
+import { Box, Text } from '..';
+import README from './README.mdx';
+import { ButtonLink, ButtonLinkSize } from '.';
+
+export default {
+  title: 'Components/ComponentLibrary/ButtonLink',
+
```
Comment (georgewrmarshall) [inline ui/components/component-library/button-link/button-link.stories.tsx:17]: nit remove line break
```suggestion
```

File: ui/components/component-library/button-link/README.mdx [hunk 3]
```diff
@@ -23,52 +23,49 @@ The `ButtonLink` accepts all [ButtonBase](/docs/components-componentlibrary-butt
...
+- `ButtonLinkSize.Lg` 48px
+- `ButtonLinkSize.Inherit` inherits the font-size of the parent element. Used for inline links in paragraphs.
 
 <Canvas>
   <Story id="components-componentlibrary-buttonlink--size-story" />
 </Canvas>
 
 ```jsx
-import { Size } from '../../../helpers/constants/design-system';
 import { ButtonLink, Text, TextVariant } from '../../component-library';
```
Comment (georgewrmarshall) [inline ui/components/component-library/button-link/README.mdx:41]: Add `ButtonLink` size here 
```suggestion
import { ButtonLink, ButtonLinkSize, Text, TextVariant } from '../../component-library';
```

File: ui/components/component-library/button-link/README.mdx [hunk 4]
```diff
@@ -133,16 +130,16 @@ import { ButtonLink } from '../../component-library';
 <ButtonLink paddingLeft={4} paddingRight={4}>
```
Comment (georgewrmarshall) [inline ui/components/component-library/button-link/README.mdx:130]: Add ButtonLink size to import in line `128`
```
import { ButtonLink, ButtonLinkSize } from '../../component-library';
```

File: ui/components/component-library/button-link/README.mdx [hunk 5]
```diff
@@ -13,7 +13,7 @@ The `ButtonLink` is an extension of `ButtonBase` to support link styles
 
 ## Props
 
-The `ButtonLink` accepts all props below as well as all [Box](/docs/components-ui-box--default-story#props) and [ButtonBase](/docs/components-componentlibrary-buttonbase--default-story#props) component props
+The `ButtonLink` accepts all props below as well as all [Box](/docs/components-componentlibrary-box--docs#props) and [ButtonBase](/docs/components-componentlibrary-buttonbase--default-story#props) component props
```
Comment (georgewrmarshall) [inline ui/components/component-library/button-link/README.mdx:16]: Let's remove this line altogether now that the props table shows props for the `Box` and `BaseButton`. We can also remove lines `20 - 22`

```suggestion
```

```
The `ButtonLink` accepts all [ButtonBase](/docs/components-componentlibrary-buttonbase--default-story#props) component props
<ArgsTable of={ButtonBase} />
```

**Review by georgewrmarshall — COMMENTED** [review]
Looking great! Some small updates to docs but nothing major
- checked ButtonLink docs, stories and controls
- checked `Button` docs, stories and controls

One think I noticed that was possibly introduced in the `ButtonBase` migration was that the loading icon in button is iconDefault. Not a big one but it may need updated in another PR

Screenshot 2023-08-03 at 2 32 45 PM
Screenshot 2023-08-03 at 2 52 54 PM

**Review by georgewrmarshall — DISMISSED** [review]
🔥 🚀

**Review by hmalik88 — APPROVED** [review]

**Review by darkwing — APPROVED** [review]

**Pull request merged**